### PR TITLE
#10289: Cursor position is shifted incorrectly in Localized number field component

### PR DESCRIPTION
**Pull request by mahmoudadel54**
## Description
In this PR, fixing cursor position shifting incorrectly in Localized number field component

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue
#10289 

**What is the current behavior?**
#10289 

**What is the new behavior?**
For Localized number field component, if user edits/adds/deletes numbers into the input, the cursor will move in a correct way.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Review by dsuren1 — CHANGES_REQUESTED** [review]
@mahmoudadel54 
In most cases, the cursor position behaves correctly, except for few these. Kindly check. Thanks!

- With fractional/decimal digit, on focus field value becomes unformatted

  https://github.com/geosolutions-it/MapStore2/assets/26929983/5afe0045-201f-47bd-b64d-f7b1211e4c0c

- Follow up to above issue, the cursor position is incorrect when deleting fractional digits probably caused by unformatted value

  https://github.com/geosolutions-it/MapStore2/assets/26929983/374bf2d1-2186-4134-b397-a1a48814a9be

**Comment by mahmoudadel54** [comment]
> @mahmoudadel54 In most cases, the cursor position behaves correctly, except for few these. Kindly check. Thanks!
> 
> * With fractional/decimal digit, on focus field value becomes unformatted
>   
>     
>       
>       
>   
>       fractional_unformatted.mp4
> * Follow up to above issue, the cursor position is incorrect when deleting fractional digits probably caused by unformatted value
>   
>     
>       
>       
>   
>       unformatted_cursor_position_incorrect.mp4

These are because this regex which located in NumericInput: 
const RE_INCOMPLETE_NUMBER = /^([+-]|[+-]?\d+\.\d+0*|[+-]\.0*|[+-]?\d+\.)?$/;
it sees number like: 321.235 as incomplete number

**Review by dsuren1 — CHANGES_REQUESTED** [review]
@mahmoudadel54 
- Add unit tests for all these cases
- Cursor position shift incorrectly as we type. I didn't see this before previous commit. Kindly check it. Thanks!
 

  https://github.com/geosolutions-it/MapStore2/assets/26929983/1dad6be4-7cd2-41e7-87ff-012caf3aa90f

**Comment by mahmoudadel54** [comment]
> @mahmoudadel54
> 
> * Add unit tests for these cases
> * Cursor position shift incorrectly as we type. I didn't see this before previous commit. Kindly check it. Thanks!
>   
>     
>       
>       
>   
>       incorrect_cursor.mp4

Hi @dsuren1 
can you please double check again? as I made double check and it works well 


https://github.com/geosolutions-it/MapStore2/assets/58145645/53efc3b9-36a0-4d5f-afbd-e834176e6318

**Comment by dsuren1** [comment]
> Hi @dsuren1 can you please double check again? as I made double check and it works well

Like I mentioned over chat, it happens when user types the value a little fast. Kindly give it a try like we discussed in chat. Thanks!

**Review by dsuren1 — CHANGES_REQUESTED** [review]
@mahmoudadel54 
Before reviewing this new change, I would also like this component to support `onKeyUp` function from consuming component in tandem with the changes incorporated for handing cursor

**Comment by mahmoudadel54** [comment]
> @mahmoudadel54 Before reviewing this new change, I would also like this component to support `onKeyUp` function from consuming component in tandem with the changes incorporated for handing cursor

Done

File: web/client/components/I18N/IntlNumberFormControl.jsx [hunk 1]
```diff
@@ -30,15 +29,69 @@ class IntlNumberFormControl extends React.Component {
...
+        onKeyUp: PropTypes.func
     }
-
     static contextTypes = {
         intl: PropTypes.object
     };
+    static defaultProps = {
+        onKeyDown: () => {},
+        onKeyUp: () => {}
+    }
```
Comment (dsuren1) [inline web/client/components/I18N/IntlNumberFormControl.jsx:42]:
```suggestion
```
Not needed

File: web/client/components/I18N/IntlNumberFormControl.jsx [hunk 2]
```diff
@@ -30,15 +29,69 @@ class IntlNumberFormControl extends React.Component {
...
+                } else if (prevValueLength - currentValueLength === 2 && groupSeparatorPrevValue > groupSeparatorCurrentValue) {    // in deleting numbers and a group separator will be reduced due to deletion
+                    this.state.inputRef.setSelectionRange(currentCursorPos - 2 || 0, currentCursorPos - 2 || 0);
+                } else if (prevValueLength - currentValueLength === 1 && groupSeparatorPrevValue === groupSeparatorCurrentValue) {  // in deleting numbers and the group separators is the same after deletion
+                    this.state.inputRef.setSelectionRange(currentCursorPos - 1 || 0, currentCursorPos - 1 || 0);
+                }
+            }
+        }
+    }
+    onKeyUp = (ev) =>{
+        this.props.onKeyUp && this.props.onKeyUp();
```
Comment (dsuren1) [inline web/client/components/I18N/IntlNumberFormControl.jsx:83]:
```suggestion
        this.props.onKeyUp && this.props.onKeyUp(ev);
```

File: web/client/components/I18N/IntlNumberFormControl.jsx [hunk 3]
```diff
@@ -30,15 +29,69 @@ class IntlNumberFormControl extends React.Component {
...
+        const currentCursorPos = this.currentInputCursor;
+        let isDelete = ev.keyCode === 8 || ev.keyCode === 46;       // delete by delete key or backspace key
+        // move the cursor of adding number at the end of input
+        if (![37, 39, 17].includes(ev.keyCode) && !isDelete && currentCursorPos === ev.target.value.length - 1) {
+            return ev.target.setSelectionRange(-1, -1);
+        }
+        return false;
```
Comment (dsuren1) [inline web/client/components/I18N/IntlNumberFormControl.jsx:90]:
```suggestion
        if (![37, 39, 17].includes(ev.keyCode) && !isDelete && currentCursorPos === ev.target.value.length - 1) {
            ev.target.setSelectionRange(-1, -1);
        }
```

**Review by dsuren1 — APPROVED** [review]

**Comment by dsuren1** [comment]
@ElenaGallo Kindly test it in DEV and let us know if it's good to be backported. Thanks!

**Comment by ElenaGallo** [comment]
Test passed, @mahmoudadel54 please backport to 2024.01.xx. Thanks

**Comment by mahmoudadel54** [comment]
> Test passed, @mahmoudadel54 please backport to 2024.01.xx. Thanks

Backport is done ---> https://github.com/geosolutions-it/MapStore2/pull/10346

**Pull request merged**